certbot/certbot/tests/util.py

```diff
@@ -149,43 +148,71 @@ def make_lineage(config_dir, testfile, ec=False):
     return conf_path
 
 
-def patch_get_utility(target='zope.component.getUtility'):
-    """Patch zope.component.getUtility to use a special mock IDisplay.
+def patch_display_service():
+    """Patch CertbotService display property to use a special mock IDisplay.
 
     The mock IDisplay works like a regular mock object, except it also
     also asserts that methods are called with valid arguments.
 
-    :param str target: path to patch
-
-    :returns: mock zope.component.getUtility
+    :returns: mock certbot.display.util._get_display
     :rtype: mock.MagicMock
 
     """
-    return mock.patch(target, new_callable=_create_get_utility_mock)
+    return mock.patch('certbot.display.util._get_display',
+                      new_callable=_create_display_service_mock)
 
 
-def patch_get_utility_with_stdout(target='zope.component.getUtility',
-                                  stdout=None):
-    """Patch zope.component.getUtility to use a special mock IDisplay.
+def patch_display_service_with_stdout(stdout=None):
+    """Patch CertbotService display property to use a special mock IDisplay.
 
     The mock IDisplay works like a regular mock object, except it also
     also asserts that methods are called with valid arguments.
 
     The `message` argument passed to the IDisplay methods is passed to
     stdout's write method.
 
-    :param str target: path to patch
     :param object stdout: object to write standard output to; it is
         expected to have a `write` method
 
-    :returns: mock zope.component.getUtility
+    :returns: mock certbot.display.util._get_display
     :rtype: mock.MagicMock
 
     """
     stdout = stdout if stdout else io.StringIO()
 
-    freezable_mock = _create_get_utility_mock_with_stdout(stdout)
-    return mock.patch(target, new=freezable_mock)
+    return mock.patch('certbot.display.util._get_display',
```

Comment:
I think I was somewhat misleading in my comment at https://github.com/certbot/certbot/pull/8835#discussion_r644970405 so I'm sorry for that.

To be clearer, I'm totally fine with us messing with internals more than we should temporarily, but I think we shouldn't add any code that does that to our public API. Since I don't think we should be messing with internals unless absolutely necessary long term, I'd like to avoid that becoming part of our public API making it harder for us to change.

I'm open to other options, but I personally think the best option for dealing with this is to just bite the bullet and fix things up to patch `display.util` now. I think the main trouble with that will be that the mocks are used slightly differently (e.g. `patched_get_utility().notification` vs `patched_display_util.notification`). If this is done, I'd also like to suggest naming these functions something like `patch_display_util*`.

Alternatively, I think there's a few ways we could temporarily patch `_get_display` in our tests without making code that does it part of our public API, but I don't think we should use internals like that between our packages and my suspicion is that after making all the changes necessary to make things like pylint happy, it'll be almost as much work as just using `display.util` now.

I know this comment is suggesting a fair bit of work, but I hope you understand my concerns here. I'm certainly open to other ideas if you have them.

Reply:
No this is on me, I forgot that this module is public.

So I am fine with patching the public part of the module, but I am not sure about the best way to do that. If we want to be able to call something like `mock_util.notification`, we need to mock the entire `display.util`module. It does not seem to be a very common approach with mocks.

Or we patch the public methods themselves, and work with their mock directly. It seems better, because the mock is more accurate. In this case, do we really want to clutter the public API of the `test` module with several `patch_get_utility*`functions? It seems easier to directly call `mock.match("certbot.display.util.method")` in the relevant tests.

Reply:
I don't think patching things at the module level is too horrible, but I agree it's a somewhat strange things to do. I also agree it'd probably make the most sense to not add more to our public test API and just patch things as needed in the relevant tests, but that seems like a ton of work and will result in a large PR that does a lot more than just find and replaces a few functions.

I'm honestly struggling a bit to come up with a plan I like here so feel free to propose an alternative, but what about this:

1. In this PR for now, the `get_display` and `set_display` functions continue to use `zope.component`. This is just a temporary measure to simplify the transition so `patch_get_utility*` continues to work on the new functions.
2. In subsequent PR(s), we move our tests away from using `patch_get_utility*`. I think just patching things in the relevant tests makes a lot of sense, however, I also think it may be worth us considering doing something with pytest like I suggested in my parenthetical [here](https://github.com/certbot/certbot/pull/8835#discussion_r644280409).
3. Once we've transitioned away from `patch_get_utility*`, we deprecate them and move the implementations of `get_display` and `set_display` away from zope to what you have here.

Reply:
I was trying to imagine some workarounds involving to monkey patch `zope.component.getUtility` but sadly I do not think it will be significantly simpler that what you propose here.

Last call to see if it gives you some idea, and if not, I will implement you proposition.

Reply:
In an ideal world where the resources on this project were infinite, I personally like my proposition in my last post the best. It results in the design I like best in the long term without adding any cruft to our API.

Since that's not the world we live in though and refactoring all the test code around IDisplay seems like a fair bit of work and significant deviation from our initial high level goal here of adding mypy type annotations, I'm beginning to think that what you have here isn't so bad as a temporary solution. I'd ideally like to essentially implement what we discussed above in the long term, but if temporarily adding one or two functions to our public test API allows us to more easily move forward here, maybe that approach isn't so bad. I'm not sure (m)any people other than us are using this API anyway.

If you want to stick with the approach you currently have, I'd suggest:

1. Maybe renaming the functions to something like `patch_display_util*` since I think "service" is the old name.
2. Make an issue briefly explaining the test refactoring that should be done here maybe including a link to this comment thread.
3. Maybe update the warning raised out of `patch_get_utility*` to say something like "Decorator certbot.tests.util.patch_get_utility is deprecated. You should now patch certbot.display.util yourself directly or use `<new test util function name>` as a temporary workaround."
4. Update the docstring of the new functions to say something like:
```
    """Patch certbot.display.util to use a special mock IDisplay.

    The mock IDisplay works like a regular mock object, except it also
    also asserts that methods are called with valid arguments.

    The mock created by this patch mocks out Certbot internals so this can be
    used like the old patch_get_utility function. That is, the mock object will
    be called by the certbot.display.util functions and the mock returned by
    that call will be used as the IDisplay object. This was done to simplify
    the transition from zope.component and mocking certbot.display.util
    functions directly in test code should be preferred over using this
    function in the future.

    :returns: patch on the function used internally by certbot.display.util to
        get an IDisplay object
    :rtype: unittest.mock._patch

    """
```
What do you think? Either of the options described here or in my prior post work for me and I'm happy to try to help implement either if you like.

Reply:
Indeed this is a realistic approach.

I made the modifications about the function names and docstring, and created the issue https://github.com/certbot/certbot/issues/8948 to keep track of the future refactoring.